core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala

```diff
@@ -0,0 +1,156 @@
+
+    val metadataResponse2 = sendMetadataRequest(allMetadataRequest, 1)
+    val controllerServer2 = servers.find(_.kafkaController.isActive()).get
+    val controllerId2 = controllerServer2.apis.brokerId
+
+    assertNotEquals("Controller id should switch to a new broker", controllerId, controllerId2)
+    assertEquals("Controller id should match the active controller after failover",
+      controllerServer.apis.brokerId, metadataResponse.controller.id)
+  }
+
+  @Test
+  def testRack() {
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    // Validate rack matches whats set in generateConfigs() above
+    metadataResponse.brokers().asScala.foreach { broker =>
+      assertEquals("Rack information should match config", s"rack/${broker.id}", broker.rack())
+    }
+  }
+
+  @Test
+  def testIsInternal() {
+    val internalTopic = TopicConstants.GROUP_METADATA_TOPIC_NAME
+    val notInternalTopic = "notInternal"
+    // create the topics
+    TestUtils.createTopic(zkUtils, internalTopic, 3, 2, servers)
+    TestUtils.createTopic(zkUtils, notInternalTopic, 3, 2, servers)
+
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    assertTrue("Response should have no errors", metadataResponse.errors().isEmpty)
+
+    val topicMetadata = metadataResponse.topicMetadata().asScala
+    val internalTopicMetadata = topicMetadata.find(_.topic == internalTopic).get
+    val notInternalTopicMetadata = topicMetadata.find(_.topic == notInternalTopic).get
+
+    assertTrue("internalTopic should show isInternal", internalTopicMetadata.isInternal)
+    assertFalse("notInternalTopic topic not should show isInternal", notInternalTopicMetadata.isInternal)
+  }
+
+  @Test
+  def testNoTopicsRequest() {
+    // create some topics
+    TestUtils.createTopic(zkUtils, "t1", 3, 2, servers)
+    TestUtils.createTopic(zkUtils, "t2", 3, 2, servers)
+
+    val metadataResponse = sendMetadataRequest(noMetadataRequest, 1)
+    assertTrue("Response should have no errors", metadataResponse.errors().isEmpty)
+    assertTrue("Response should have no topics", metadataResponse.topicMetadata().isEmpty)
+  }
+
+  @Test
+  def testReplicaDownResponse() {
+    val replicaDownTopic = "replicaDown"
+    val replicaCount = 3
+
+    // create a topic with 3 replicas
+    TestUtils.createTopic(zkUtils, replicaDownTopic, 1, replicaCount, servers)
+
+    // Kill a replica node that is not the leader
+    val metadataResponse = sendMetadataRequest(new MetadataRequest(List(replicaDownTopic).asJava), 1)
+    val partitionMetadata = metadataResponse.topicMetadata().asScala.head.partitionMetadata().asScala.head
+    val downNode = servers.find { server =>
+      val serverId = server.apis.brokerId
+      val leaderId = partitionMetadata.leader.id()
+      val replicaIds = partitionMetadata.replicas().asScala.map(_.id())
+      serverId != leaderId && replicaIds.contains(serverId)
+    }.get
+    downNode.shutdown()
+
+    TestUtils.waitUntilTrue({ () =>
```

Comment:
Nitpick: for cases like this, I think it's more readable to have the opening brace after `=>`. For cases where the opening brace can replace the opening parenthesis, it makes sense to position the brace like this case. A bit subjective though.